Remove recursion in merkle tree implementation

howjmay opened the pull request:
Remove the recursion in merkle tree in merkle.c

Fixes #15 

The memory depth is linealy propontional to the depth of merkle tree in the original recursive version of binary_tree_size(). but the memory dpeth is constant in the optimized version of binary_tree_size(). Moreover, the optimized version only calculates once instead of using iteration.

### memory reduction
#### binary_tree_size()
The original binary_tree_size()'s stack memory size is "32 * tree_depth" bytes in which each frame size of binary_tress size() is 32 bytes.

The opitimized binary_tree_size()'s stack memory size is 64 bytes.

##### memory reduction ratio = 64/(32 *tree_depth)
#### merkle_node_index_traverse()
The original merkle_node_index_traverse()'s stack memory size is "80 * tree_depth" bytes in which each frame size of merkle_node_index_traverse() is 50 bytes.

The opitimized merkle_node_index_traverse()'s stack memory size is 64 bytes.

##### memory reduction ratio = 64/(80 *tree_depth)



### Performance 
Moreover, I replaced the function call, binary_tree_size(), in
merkle_node_index_traverse() with calculating directly.


#### binary_tree_size()
binary_tree_size() testing result for calculating a merkle tree
whose tree_depth is 20 and acc = 99 for 10000 times
=====================testing result==========================
    recursive one's cost clocks = 5414
    optimized one's cost clocks = 185

#### merkle_node_index_traverse()
merkle_node_index_traverse() testing result for calculating all the
nodes's indexes of a 20 level merkle tree
=====================testing result==========================
    recursive one's cost clocks = 318673
    optimized one's cost clocks = 185648

# Test Plan:
use test_merkle.c in mam/v1/tests

jserv commented on mam/v1/merkle.c:
```diff
@@ -28,7 +25,7 @@ size_t merkle_size(size_t const leaf_count) {
 size_t merkle_depth(size_t const node_count) {
   size_t depth = 0;
-
+  
```
Avoid such non-functioned change.

howjmay replied:
thanks. I fixed it

howjmay committed:
Remove recursion from merkle.c

Remove the recursion in merkle tree in merkle.c

jserv commented on mam/v1/merkle.c:
```diff
@@ -28,7 +25,6 @@ size_t merkle_size(size_t const leaf_count) {
 
 size_t merkle_depth(size_t const node_count) {
   size_t depth = 0;
-
```
Don't change this line.

jserv commented on mam/v1/merkle.c:
```diff
@@ -38,15 +34,26 @@ size_t merkle_depth(size_t const node_count) {
+  int tree_depth_width_divide_scale = 1;
+  int index = 0;
+  int width_cursor = width;
+  int tree_depth_width = 1 << depth;
```
`tree_depth_width` might not be a meaningful name. Can you improve it?

howjmay committed:
change varible name

Change variable name from tree_depth_width and
tree_depth_width_divide_scale to width_of_leaves and
width_of_leaves_divide_scale respectively.

jserv commented:
The commit messages should mention the following:
1. performance gain after switching to iteration-based implementation;
2. memory reduction ration because of no extra stack allocation;

Please use `git rebase -i` to rework message(s).

howjmay committed:
Performance optimization

Remove the recursion in merkle tree in merkle.c

The memory depth is linealy propontional to the depth of merkle tree in the original recursive version of binary_tree_size(). but the memory dpeth is constant in the optimized version of binary_tree_size(). Moreover, the optimized version only calculates once instead of using iteration.

memory reduction
===========binary_tree_size()===========
The original binary_tree_size()'s stack memory size is "32 * tree_depth" bytes in which each frame size of binary_tress size() is 32 bytes.

The opitimized binary_tree_size()'s stack memory size is 64 bytes.

memory reduction ratio = 64/(32 *tree_depth)
===========merkle_node_index_traverse()===========
The original merkle_node_index_traverse()'s stack memory size is "80 * tree_depth" bytes in which each frame size of merkle_node_index_traverse() is 50 bytes.

The opitimized merkle_node_index_traverse()'s stack memory size is 64 bytes.

memory reduction ratio = 64/(80 *tree_depth)

Moreover, I replaced the function call, binary_tree_size(), in
merkle_node_index_traverse() with calculating directly.

merkle_node_index_traverse() testing result for calculating all the
nodes's indexes of a 20 level merkle tree
=====================testing result==========================
    recursive one's cost clocks = 318673
    optimized one's cost clocks = 185648

binary_tree_size() testing result for calculating a merkle tree
whose tree_depth is 20 and acc = 99 for 10000 times
=====================testing result==========================
    recursive one's cost clocks = 5414
    optimized one's cost clocks = 185

jserv commented:
It would be explicit to change subject to `Remove recursion in merkle tree implementation".

howjmay commented:
@thibault-martinez ?

thibault-martinez commented:
Hi @HowJMay, thanks for this PR, I'll review it ASAP.
Great perf analysis by the way !

howjmay commented:
thank you><

howjmay commented:
@thibault-martinez Hi! 
It has been a long time since I sent this PR.
I wonder the status of code review of this PR.
I really hope that my PR can be merged into this porject. Thanks !

thibault-martinez commented:
Hi @HowJMay , could you solve conflicts on this one please ?

thibault-martinez commented:
Hi @HowJMay , could you solve conflict on this please ?

thibault-martinez commented:
Hi @HowJMay , could you solve conflict on this please ?

thibault-martinez commented:
Hi @HowJMay , could you solve conflict on this PR please ?

thibault-martinez commented:
Sorry if you received my comment multiple times, GitHub is having difficulties today.
Your implementation of `merkle_node_index_traverse` gives wrong results for two reasons.
I'll point them out in the review with some other little things.

thibault-martinez commented on mam/v1/merkle.c:
```diff
@@ -38,15 +35,26 @@ size_t merkle_depth(size_t const node_count) {
 static size_t merkle_node_index_traverse(size_t const acc, size_t const depth,
```
You are not using `acc` at all.

thibault-martinez replied:
You are not using `acc` at all.

thibault-martinez commented on mam/v1/merkle.c:
```diff
@@ -38,15 +35,26 @@ size_t merkle_depth(size_t const node_count) {
+    return 0;
+
+  int depth_cursor = 1;
+  int width_of_leaves_divide_scale = 1;
```
You don't need this variable, just use `depth_cursor`.

thibault-martinez commented on mam/v1/merkle.c:
```diff
@@ -38,15 +35,26 @@ size_t merkle_depth(size_t const node_count) {
+
+  int depth_cursor = 1;
+  int width_of_leaves_divide_scale = 1;
+  int index = 0;
```
Initialize it at `depth + acc` instead of adding `depth` at the end.

thibault-martinez commented on mam/v1/merkle.c:
```diff
@@ -38,15 +35,26 @@ size_t merkle_depth(size_t const node_count) {
+
+  while (depth_cursor <= depth) {
+    if (width_cursor >= (width_of_leaves >> width_of_leaves_divide_scale)) {
+      index += ((1 << (depth + 1)) - 1);
```
You are always adding the same tree here

thibault-martinez commented on mam/v1/merkle.c:
```diff
@@ -38,15 +35,26 @@ size_t merkle_depth(size_t const node_count) {
+  int width_of_leaves = 1 << depth;
+
+  while (depth_cursor <= depth) {
+    if (width_cursor >= (width_of_leaves >> width_of_leaves_divide_scale)) {
```
Instead of shifting `1` to the left then to the right, just shift `1` to the left of `depth - depth_cursor`

howjmay commented:
I am really sorry for that I did some really stupid error in the last version. 
I have fixed them out. Now, it can run successfully.

thibault-martinez commented:
No worries, that's what reviews are for.
Just some final nits and it'll be good to merge.
Please solve conflicts as well.

thibault-martinez commented on mam/v1/merkle.c:
```diff
@@ -35,18 +32,33 @@ size_t merkle_depth(size_t const node_count) {
+  int depth_cursor = 1;
+  int index = depth + acc;
+  int width_cursor = width;
+  int width_of_leaves = 1 << depth;
```
Use `size_t` for these variables.

thibault-martinez commented on mam/v1/merkle.c:
```diff
@@ -35,18 +32,33 @@ size_t merkle_depth(size_t const node_count) {
-    return merkle_node_index_traverse(
-        1 + acc + binary_tree_size(0, tree_depth - 1), depth - 1,
-        width - (1 << (depth - 1)), tree_depth - 1);
+  if (!tree_depth) return 0;
```
Put braces around the `if`

thibault-martinez commented:
@HowJMay ,
Thank you for this really great PR !
Please rebase to develop and it'll be good to merge.

howjmay commented:
thank you for reviewing :)

jserv commented:
@HowJMay, Don't use `git merge`. Instead, use `git rebase`: https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request

howjmay committed:
Fix errors

thibault-martinez commented on utils/merkle.c:
```diff
@@ -1,22 +1,12 @@
- *
- * Refer to the LICENSE file for licensing information
- */
-
```
Can you bring back the copyright ?

thibault-martinez commented on utils/merkle.c:
```diff
@@ -1,22 +1,12 @@
 
 #include "common/sign/v2/iss_curl.h"
-#include "utils/merkle.h"
+#include "mam/v1/merkle.h"
```
Wrong path

thibault-martinez commented on utils/merkle.c:
```diff
@@ -1,22 +1,12 @@
+#include "mam/v1/merkle.h"
 
-static trit_t const merkle_null_hash[HASH_LENGTH_TRIT] = {0};
+static trit_t const merkle_null_hash[HASH_LENGTH] = {0};
```
`HASH_LENGTH` does not exist

howjmay replied:
OK I will fix them this weekend sorry

howjmay committed:
Fix rebase errors

thibault-martinez committed:
Merge branch 'develop' into mam

thibault-martinez committed:
Merge branch 'develop' into mam

thibault-martinez commented:
@HowJMay Thanks again for this great PR !
